File: code/summarize_checkins.R
# summarize_checkins.R
# Reduces loc-brightkite_totalCheckins.txt (~4.5M rows) to a small per-user
# homes file (~1 MB) that the CoLaS pipeline accepts directly.
#
# Run from the folder that contains the check-ins file, e.g.:
#   cd /Users/mariospapamichalis/Downloads
#   Rscript summarize_checkins.R
#
# Output: brightkite_homes.csv.gz  with columns
#   user, med_lat, med_lon, n_valid, n_checkins
# (median home over valid coordinates; n_valid = valid check-ins,
#  n_checkins = all check-in rows -- matching the pipeline's semantics).

infile  <- "loc-brightkite_totalCheckins.txt"
outfile <- "brightkite_homes.csv.gz"

if (requireNamespace("data.table", quietly = TRUE)) {
  library(data.table)
  ck <- fread(infile, header = FALSE, sep = "\t",
              col.names = c("user", "time", "lat", "lon", "loc_id"),
              select = c(1, 3, 4),
              colClasses = list(integer = 1, numeric = c(3, 4)))
  setnames(ck, c("user", "lat", "lon"))
  cnt_all <- ck[, .(n_checkins = .N), by = user]
  ok <- ck[is.finite(lat) & is.finite(lon) &
           !(lat == 0 & lon == 0) &
           lat >= -90 & lat <= 90 & lon >= -180 & lon <= 180]
  homes <- ok[, .(med_lat = median(lat), med_lon = median(lon),
                  n_valid = .N), by = user]
  homes <- merge(homes, cnt_all, by = "user")
  setorder(homes, user)
  fwrite(homes, outfile)   # .gz suffix -> gzip automatically
  cat(sprintf("wrote %s: %d users\n", outfile, nrow(homes)))
} else {
  # base-R fallback (slower, a few minutes; needs ~2 GB RAM)
  ck <- read.table(infile, sep = "\t", header = FALSE, quote = "",
                   comment.char = "",
                   colClasses = c("integer", "NULL", "numeric",
                                  "numeric", "NULL"))
  names(ck) <- c("user", "lat", "lon")
  n_checkins <- table(ck$user)
  ok <- ck[is.finite(ck$lat) & is.finite(ck$lon) &
           !(ck$lat == 0 & ck$lon == 0) &
           ck$lat >= -90 & ck$lat <= 90 &
           ck$lon >= -180 & ck$lon <= 180, ]
  agg <- aggregate(cbind(lat, lon) ~ user, data = ok, FUN = median)
  n_valid <- table(ok$user)
  agg$n_valid <- as.integer(n_valid[as.character(agg$user)])
  agg$n_checkins <- as.integer(n_checkins[as.character(agg$user)])
  names(agg) <- c("user", "med_lat", "med_lon", "n_valid", "n_checkins")
  agg <- agg[order(agg$user), ]
  con <- gzfile(outfile, "w")
  write.csv(agg, con, row.names = FALSE)
  close(con)
  cat(sprintf("wrote %s: %d users\n", outfile, nrow(agg)))
}

File: code/summarize_checkins.py
#!/usr/bin/env python3
"""Pure-stdlib alternative to summarize_checkins.R (no packages needed).

  cd /Users/mariospapamichalis/Downloads
  python3 summarize_checkins.py loc-brightkite_totalCheckins.txt

Accepts the .txt or the original .txt.gz.  Writes brightkite_homes.csv.gz
with columns user, med_lat, med_lon, n_valid, n_checkins.
"""
import sys, gzip, csv
from collections import defaultdict
from statistics import median

import os, re
src = sys.argv[1] if len(sys.argv) > 1 else "loc-brightkite_totalCheckins.txt"
if len(sys.argv) > 2:
    out_name = sys.argv[2]
else:
    base = os.path.basename(src)
    m = re.match(r"loc[-_]?([A-Za-z]+)", base)
    out_name = (m.group(1).lower() if m else "network") + "_homes.csv.gz"
opener = gzip.open if src.endswith(".gz") else open

lats = defaultdict(list); lons = defaultdict(list); n_all = defaultdict(int)
with opener(src, "rt") as fh:
    for line in fh:
        p = line.rstrip("\n").split("\t")
        if len(p) < 5:
            continue
        try:
            u = int(p[0]); la = float(p[2]); lo = float(p[3])
        except ValueError:
            continue
        n_all[u] += 1
        if (la == 0.0 and lo == 0.0) or not (-90 <= la <= 90 and -180 <= lo <= 180):
            continue
        lats[u].append(la); lons[u].append(lo)

with gzip.open(out_name, "wt", newline="") as fh:
    w = csv.writer(fh)
    w.writerow(["user", "med_lat", "med_lon", "n_valid", "n_checkins"])
    for u in sorted(lats):
        w.writerow([u, f"{median(lats[u]):.6f}", f"{median(lons[u]):.6f}",
                    len(lats[u]), n_all[u]])
print(f"wrote {out_name}: {len(lats)} users")
